signer-cloud-server/src/main/java/com/wultra/signercloud/server/document/DocumentRangesService.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Signer Cloud
+ * Copyright (C) 2025 Wultra s.r.o.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.wultra.signercloud.server.document;
+
+import org.springframework.http.HttpRange;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+/**
+ * Service for parsing HTTP {@code Range} header and merging overlapping ranges.
+ *
+ * @author Michal Rozehnal, michal.rozehnal@wultra.com
+ */
+@Service
+class DocumentRangesService {
+
+    /**
+     * Parse HTTP {@code Range} header from String to {@link  List<DocumentPart>}.
+     *
+     * Parts are merged if they overlap or are adjacent. They are returned in the same order as they were specified in the input.
+     *
+     * @param ranges ranges header value
+     * @param fileSize size of the file to which the ranges apply
+     * @return list of ranges
+     */
+    List<DocumentPart> getParts(final String ranges, final long fileSize) {
+        final var parsedRanges = parseRanges(ranges);
+
+        final var parts = new ArrayList<DocumentPart>(parsedRanges.size());
+
+        for (int i = 0; i < parsedRanges.size(); i++) {
+            final var parsedRange = parsedRanges.get(i);
+            final var start = (int) parsedRange.getRangeStart(fileSize);
+            final var end = (int) parsedRange.getRangeEnd(fileSize);
+
+            parts.add(new DocumentPart(start, end, i));
+        }
+
+        return mergeParts(parts);
+    }
+
+    private static List<HttpRange> parseRanges(final String ranges) {
+        try {
+            return HttpRange.parseRanges(ranges);
+        } catch (final IllegalArgumentException e) {
+            final var message = String.format("Invalid range header: %s Reason: %s", ranges, e.getMessage());
```

Comment:
Nitpick, you may write
```suggestion
            final var message = "Invalid range header: %s Reason: %s".formatted(ranges, e.getMessage());
```

Reply:
Right, this is more readable. Fixed